### PR TITLE
Deprecate `GatewayClient`

### LINKED ISSUE #919: Pending Deprecate GatewayClient
### Feature Request

GatewayClient should become deprecated in favour of FullNodeClient.

### PR DESCRIPTION
**Pull request by tomek0123456789**
Closes #919

## Introduced changes

- `GatewayClient` is now deprecated and throws a warning whenever it is created

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/net/gateway_client.py [hunk 1]
```diff
@@ -67,13 +67,21 @@ def __init__(
...
         Client for interacting with Starknet gateway.
 
         :param net: Target network for the client. Can be a string with URL, one of ``"mainnet"``, ``"testnet"``
                     or dict with ``"feeder_gateway_url"`` and ``"gateway_url"`` fields
         :param session: Aiohttp session to be used for request. If not provided, client will create a session for
                         every request. When using a custom session, user is responsible for closing it manually.
         """
+        warnings.warn(
+            "GatewayClient is deprecated and will not be supported in the future. Please use FullNodeClient instead.",
+            DeprecationWarning,
```
Comment (drknzz) [inline starknet_py/net/gateway_client.py:82]: Let's make it a pending deprecation

File: starknet_py/net/gateway_client.py [hunk 2]
```diff
@@ -67,13 +67,21 @@ def __init__(
         session: Optional[aiohttp.ClientSession] = None,
     ):
         """
+        .. deprecated:: 0.15.0
```
Comment (drknzz) [inline starknet_py/net/gateway_client.py:70]: The next version is 0.17.1 🌵

File: starknet_py/net/gateway_client.py [hunk 3]
```diff
@@ -67,13 +67,21 @@ def __init__(
...
+            GatewayClient is deprecated and will not be supported in the future.
```
Comment (drknzz) [inline starknet_py/net/gateway_client.py:71]: Let's change it to something not so directly speaking about deprecation, something like _Gateway / Feeder Gateway API will become deprecated in the future. As a result, GatewayClient won't work and will eventually be removed. Consider migrating to FullNodeClient_

File: starknet_py/net/gateway_client.py [hunk 4]
```diff
@@ -67,13 +67,21 @@ def __init__(
...
+
         Client for interacting with Starknet gateway.
 
         :param net: Target network for the client. Can be a string with URL, one of ``"mainnet"``, ``"testnet"``
                     or dict with ``"feeder_gateway_url"`` and ``"gateway_url"`` fields
         :param session: Aiohttp session to be used for request. If not provided, client will create a session for
                         every request. When using a custom session, user is responsible for closing it manually.
         """
+        warnings.warn(
+            "GatewayClient is deprecated and will not be supported in the future. Please use FullNodeClient instead.",
```
Comment (drknzz) [inline starknet_py/net/gateway_client.py:81]: Something similar here

File: docs/quickstart.rst [hunk 5]
```diff
@@ -1,9 +1,27 @@
 Quickstart
 ==========
 
+Using FullNodeClient
+--------------------
+A :ref:`Client` is a facade for interacting with Starknet.
+:ref:`FullNodeClient` is a client which interacts
+with a Starknet full node like `pathfinder <https://github.com/eqlabs/pathfinder>`_.
+Like the GatewayClient, it supports read and write operations. Also as with GatewayClient,
```
Comment (drknzz) [inline docs/quickstart.rst:9]: Let's rephrase these sentences so they don't speak about GatewayClient

File: docs/quickstart.rst [hunk 6]
```diff
@@ -1,9 +1,27 @@
...
+
+You can check out all of the FullNodeClient's methods here: :ref:`FullNodeClient`.
+
 Using GatewayClient
 -------------------
-A :ref:`Client` is a facade for interacting with Starknet.
+
+.. warning::
+
+    ``GatewayClient`` is deprecated and will not be supported in the future. Please use :ref:`FullNodeClient`.
```
Comment (drknzz) [inline docs/quickstart.rst:23]: Same here as below

File: docs/quickstart.rst [hunk 7]
```diff
@@ -1,9 +1,27 @@
 Quickstart
 ==========
 
+Using FullNodeClient
+--------------------
+A :ref:`Client` is a facade for interacting with Starknet.
+:ref:`FullNodeClient` is a client which interacts
+with a Starknet full node like `pathfinder <https://github.com/eqlabs/pathfinder>`_.
```
Comment (drknzz) [inline docs/quickstart.rst:8]: Maybe we could mention more nodes?
- https://github.com/starkware-libs/papyrus
- https://github.com/NethermindEth/juno

File: starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+from starknet_py.net.full_node_client import FullNodeClient
+
+
+def test_synchronous_gateway_client():
```
Comment (drknzz) [inline starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py:4]:
```suggestion
def test_synchronous_full_node_client():
```

File: starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    # pylint: disable=no-member, unused-variable
+    # docs: start
+    node_url = "https://your.node.url"
+    synchronous_full_node_client = FullNodeClient(node_url=node_url)
+    # docs: end
+
+    synchronous_full_node_client = FullNodeClient(
+        node_url="http://188.34.188.184:9545/rpc/v0.3"
+    )
```
[the remark this answers is not in the document]
Comment (tomek0123456789) [reply starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py:13]: The problem is, when you do it this way, FullNode throws an error in constructor that it cannot connect to the url provided (no surprise), so we need to use _some_ working node url.

File: starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    # pylint: disable=import-outside-toplevel
+    from starknet_py.contract import Contract
+
+    full_node_client = fixture_client
+    key = 1234
+    contract = Contract.from_address_sync(
+        address=map_contract.address, provider=account
+    )
+    invocation = contract.functions["put"].invoke_sync(key, 7, max_fee=int(1e16))
+    invocation.wait_for_acceptance_sync()
```
Comment (drknzz) [inline starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py:22]: 👀

Comment (tomek0123456789) [reply starknet_py/tests/e2e/docs/quickstart/test_synchronous_full_node_client.py:22]: 🤫

**Review by drknzz — APPROVED** [review]
👌

**Review by drknzz — APPROVED** [review]

**Pull request merged**